File: src/com/Torvald/Terrarum/GameMap/GameMap.java
/* 
 * MapLoader version 1.2
 * Release date 2013-05-20
 * Copyright 2013 SKYHi14
 * 
 * The program is distributed in GNU GPL Licence version 3.
 * See http://www.gnu.org/licenses/gpl.html for information.
 */

package com.Torvald.Terrarum.GameMap;

import org.newdawn.slick.SlickException;

public class GameMap {

    //layers
    private volatile MapLayer layerWall;
    private volatile MapLayer layerTerrain;
    private volatile MapLayer layerWire;
    private volatile PairedMapLayer wallDamage;
    private volatile PairedMapLayer terrainDamage;

    //properties
    private int width;
    private int height;
    private int spawnX;
    private int spawnY;

    public static transient final int WALL = 0;
    public static transient final int TERRAIN = 1;
    public static transient final int WIRE = 2;

    //public World physWorld = new World( new Vec2(0, -TerrarumMain.game.gravitationalAccel) );
    //physics
    private float gravitation;
    private int globalLight;
    private WorldTime worldTime;

    public static transient final int TILES_SUPPORTED = MapLayer.RANGE * PairedMapLayer.RANGE;
    public static transient final byte BITS = 1; // 1 for Byte, 2 for Char, 4 for Int, 8 for Long
    public static transient final byte LAYERS = 4; // terrain, wall (terrainDamage + wallDamage), wire

    /**
     * @param width
     * @param height
     * @throws SlickException
     */
    public GameMap(int width, int height) throws SlickException {
        this.width = width;
        this.height = height;
        this.spawnX = width / 2;
        this.spawnY = 200;

        layerTerrain = new MapLayer(width, height);
        layerWall = new MapLayer(width, height);
        layerWire = new MapLayer(width, height);
        terrainDamage = new PairedMapLayer(width, height);
        wallDamage = new PairedMapLayer(width, height);

        globalLight = (char) 0;
        worldTime = new WorldTime();
    }

    public void setGravitation(float g) {
        gravitation = g;
    }

    /**
     * Get 2d array data of terrain
     *
     * @return byte[][] terrain layer
     */
    public byte[][] getTerrainArray() {
        return layerTerrain.data;
    }

    /**
     * Get 2d array data of wall
     *
     * @return byte[][] wall layer
     */
    public byte[][] getWallArray() {
        return layerWall.data;
    }

    /**
     * Get 2d array data of wire
     *
     * @return byte[][] wire layer
     */
    public byte[][] getWireArray() {
        return layerWire.data;
    }

    /**
     * Get paired array data of damage codes.
     * Format: 0baaaabbbb, aaaa for x = 0, 2, 4, ..., bbbb for x = 1, 3, 5, ...
     * @return byte[][] damage code pair
     */
    public byte[][] getDamageDataArray() {
        return terrainDamage.dataPair;
    }

    /**
     * Get MapLayer object of terrain
     *
     * @return MapLayer terrain layer
     */
    public MapLayer getLayerTerrain() {
        return layerTerrain;
    }

    public MapLayer getLayerWall() {
        return layerWall;
    }

    public MapLayer getLayerWire() {
        return layerWire;
    }

    public PairedMapLayer getTerrainDamage() {
        return terrainDamage;
    }

    public PairedMapLayer getWallDamage() {
        return wallDamage;
    }

    public int getTileFromWall(int x, int y) {
        return layerWall.getTile(x, y) * PairedMapLayer.RANGE + getWallDamage(x, y);
    }

    public int getTileFromTerrain(int x, int y) {
        return layerTerrain.getTile(x, y) * PairedMapLayer.RANGE + getTerrainDamage(x, y);
    }

    public int getTileFromWire(int x, int y) {
        return layerWire.getTile(x, y);
    }

    public int getWallDamage(int x, int y) {
        return wallDamage.getData(x, y);
    }

    public int getTerrainDamage(int x, int y) {
        return terrainDamage.getData(x, y);
    }

    /**
     * Set the tile of wall as specified, with damage value of zero.
     * @param x
     * @param y
     * @param combinedTilenum (tilenum * 16) + damage
     */
    public void setTileWall(int x, int y, int combinedTilenum) {
        setTileWall(x, y
                , (byte) (combinedTilenum / PairedMapLayer.RANGE)
                , combinedTilenum % PairedMapLayer.RANGE);
    }

    /**
     * Set the tile of wall as specified, with damage value of zero.
     * @param x
     * @param y
     * @param combinedTilenum (tilenum * 16) + damage
     */
    public void setTileTerrain(int x, int y, int combinedTilenum) {
        setTileTerrain(x, y
                , (byte) (combinedTilenum / PairedMapLayer.RANGE)
                , combinedTilenum % PairedMapLayer.RANGE);
    }

    public void setTileWall(int x, int y, byte tile, int damage) {
        layerWall.setTile(x, y, tile);
        wallDamage.setData(x, y, damage);
    }

    public void setTileTerrain(int x, int y, byte tile, int damage) {
        layerTerrain.setTile(x, y, tile);
        terrainDamage.setData(x, y, damage);
    }

    public void setTileWire(int x, int y, byte tile) {
        layerWire.data[y][x] = tile;
    }

    public int getTileFrom(int mode, int x, int y) {
        if (mode == TERRAIN) { return getTileFromTerrain(x, y); }
        else if (mode == WALL) { return getTileFromWall(x, y); }
        else if (mode == WIRE) { return getTileFromWire(x, y); }
        else throw new IllegalArgumentException("illegal mode input: " + String.valueOf(mode));
    }

    public void overwriteLayerWall(MapLayer layerData) {
        layerWall = layerData;
    }

    public void overwriteLayerTerrain(MapLayer layerData) {
        layerTerrain = layerData;
    }

    private int uint8ToInt32(byte x) {
        int ret;
        if ((x & 0b1000_0000) != 0) {
            ret = x & 0b1111_1111;
        } else {
            ret = x;
        }
        return ret;
    }

    public float getGravitation() {
        return gravitation;
    }

    public int getGlobalLight() {
        return globalLight;
    }

    public void setGlobalLight(int globalLight) {
        this.globalLight = globalLight;
    }

    public WorldTime getWorldTime() {
        return worldTime;
    }

    public void updateWorldTime(int delta) {
        worldTime.update(delta);
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public int getSpawnX() {
        return spawnX;
    }

    public int getSpawnY() {
        return spawnY;
    }
}

File: src/com/Torvald/Terrarum/GameMap/GameMap.kt

package com.torvald.terrarum.gamemap

import org.newdawn.slick.SlickException

class GameMap
/**
 * @param width
 * *
 * @param height
 * *
 * @throws SlickException
 */
@Throws(SlickException::class)
constructor(//properties
        val width: Int, val height: Int) {

    //layers
    val layerWall: MapLayer
    /**
     * Get MapLayer object of terrain

     * @return MapLayer terrain layer
     */
    val layerTerrain: MapLayer
    val layerWire: MapLayer
    val wallDamage: PairedMapLayer
    val terrainDamage: PairedMapLayer
    val spawnX: Int
    val spawnY: Int

    //public World physWorld = new World( new Vec2(0, -TerrarumMain.game.gravitationalAccel) );
    //physics
    var gravitation: Float = 0.toFloat()
    var globalLight: Int = 0
    val worldTime: WorldTime

    init {
        this.spawnX = width / 2
        this.spawnY = 200

        layerTerrain = MapLayer(width, height)
        layerWall = MapLayer(width, height)
        layerWire = MapLayer(width, height)
        terrainDamage = PairedMapLayer(width, height)
        wallDamage = PairedMapLayer(width, height)

        globalLight = 0.toChar().toInt()
        worldTime = WorldTime()
    }

    /**
     * Get 2d array data of terrain

     * @return byte[][] terrain layer
     */
    val terrainArray: Array<ByteArray>
        get() = layerTerrain.data

    /**
     * Get 2d array data of wall

     * @return byte[][] wall layer
     */
    val wallArray: Array<ByteArray>
        get() = layerWall.data

    /**
     * Get 2d array data of wire

     * @return byte[][] wire layer
     */
    val wireArray: Array<ByteArray>
        get() = layerWire.data

    /**
     * Get paired array data of damage codes.
     * Format: 0baaaabbbb, aaaa for x = 0, 2, 4, ..., bbbb for x = 1, 3, 5, ...
     * @return byte[][] damage code pair
     */
    val damageDataArray: Array<ByteArray>
        get() = terrainDamage.dataPair

    fun getTileFromWall(x: Int, y: Int): Int? {
        val wall: Int? = layerWall.getTile(x, y)
        val wallDamage: Int? = getWallDamage(x, y)
        return if (wall == null || wallDamage == null)
            null
        else
            wall * PairedMapLayer.RANGE + wallDamage
    }

    fun getTileFromTerrain(x: Int, y: Int): Int? {
        val terrain: Int? = layerTerrain.getTile(x, y)
        val terrainDamage: Int? = getTerrainDamage(x, y)
        return if (terrain == null || terrainDamage == null)
            null
        else
            terrain * PairedMapLayer.RANGE + terrainDamage
    }

    fun getTileFromWire(x: Int, y: Int): Int? {
        return layerWire.getTile(x, y)
    }

    fun getWallDamage(x: Int, y: Int): Int? {
        return wallDamage.getData(x, y)
    }

    fun getTerrainDamage(x: Int, y: Int): Int? {
        return terrainDamage.getData(x, y)
    }

    /**
     * Set the tile of wall as specified, with damage value of zero.
     * @param x
     * *
     * @param y
     * *
     * @param combinedTilenum (tilenum * 16) + damage
     */
    fun setTileWall(x: Int, y: Int, combinedTilenum: Int) {
        setTileWall(x, y, (combinedTilenum / PairedMapLayer.RANGE).toByte(), combinedTilenum % PairedMapLayer.RANGE)
    }

    /**
     * Set the tile of wall as specified, with damage value of zero.
     * @param x
     * *
     * @param y
     * *
     * @param combinedTilenum (tilenum * 16) + damage
     */
    fun setTileTerrain(x: Int, y: Int, combinedTilenum: Int) {
        setTileTerrain(x, y, (combinedTilenum / PairedMapLayer.RANGE).toByte(), combinedTilenum % PairedMapLayer.RANGE)
    }

    fun setTileWall(x: Int, y: Int, tile: Byte, damage: Int) {
        layerWall.setTile(x, y, tile)
        wallDamage.setData(x, y, damage)
    }

    fun setTileTerrain(x: Int, y: Int, tile: Byte, damage: Int) {
        layerTerrain.setTile(x, y, tile)
        terrainDamage.setData(x, y, damage)
    }

    fun setTileWire(x: Int, y: Int, tile: Byte) {
        layerWire.data[y][x] = tile
    }

    fun getTileFrom(mode: Int, x: Int, y: Int): Int? {
        if (mode == TERRAIN) {
            return getTileFromTerrain(x, y)
        }
        else if (mode == WALL) {
            return getTileFromWall(x, y)
        }
        else if (mode == WIRE) {
            return getTileFromWire(x, y)
        }
        else
            throw IllegalArgumentException("illegal mode input: " + mode.toString())
    }

    fun updateWorldTime(delta: Int) {
        worldTime.update(delta)
    }

    companion object {

        @Transient val WALL = 0
        @Transient val TERRAIN = 1
        @Transient val WIRE = 2

        @Transient val TILES_SUPPORTED = MapLayer.RANGE * PairedMapLayer.RANGE
        @Transient val BITS: Byte = 1 // 1 for Byte, 2 for Char, 4 for Int, 8 for Long
        @Transient val LAYERS: Byte = 4 // terrain, wall (terrainDamage + wallDamage), wire
    }
}